component/loki/source/docker/docker.go

```diff
@@ -0,0 +1,261 @@
+		},
+	})
+}
+
+const (
+	dockerLabel                = model.MetaLabelPrefix + "docker_"
+	dockerLabelContainerPrefix = dockerLabel + "container_"
+	dockerLabelContainerID     = dockerLabelContainerPrefix + "id"
+)
+
+// Arguments holds values which are used to configure the loki.source.docker
+// component.
+type Arguments struct {
+	Host         string              `river:"host,attr"`
+	Targets      []discovery.Target  `river:"targets,attr"`
+	ForwardTo    []loki.LogsReceiver `river:"forward_to,attr"`
+	Labels       map[string]string   `river:"labels,attr,optional"`
+	RelabelRules flow_relabel.Rules  `river:"relabel_rules,attr,optional"`
+}
+
+var (
+	_ component.Component      = (*Component)(nil)
+	_ component.DebugComponent = (*Component)(nil)
+)
+
+// Component implements the loki.source.file component.
+type Component struct {
+	opts    component.Options
+	metrics *dt.Metrics
+
+	mut           sync.RWMutex
+	args          Arguments
+	manager       *Manager
+	lastOptions   *Options
+	handler       loki.LogsReceiver
+	receivers     []loki.LogsReceiver
+	posFile       positions.Positions
+	rcs           []*relabel.Config
+	defaultLabels model.LabelSet
+}
+
+// New creates a new loki.source.file component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	err := os.MkdirAll(o.DataPath, 0750)
+	if err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+	positionsFile, err := positions.New(o.Logger, positions.Config{
+		SyncPeriod:        10 * time.Second,
+		PositionsFile:     filepath.Join(o.DataPath, "positions.yml"),
+		IgnoreInvalidYaml: false,
+		ReadOnly:          false,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Component{
+		opts:    o,
+		metrics: dt.NewMetrics(o.Registerer),
+
+		handler:   make(loki.LogsReceiver),
+		receivers: args.ForwardTo,
+		posFile:   positionsFile,
+	}
+
+	// Call to Update() to start readers and set receivers once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	defer c.posFile.Stop()
+
+	defer func() {
+		c.mut.Lock()
+		defer c.mut.Unlock()
+
+		// Guard for safety, but it's not possible for Run to be called without
+		// c.tailer being initialized.
+		if c.manager != nil {
+			c.manager.Stop()
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.handler:
+			for _, receiver := range c.receivers {
+				receiver <- entry
+			}
+		}
+	}
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
+	newArgs := args.(Arguments)
+
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	c.args = newArgs
+	c.receivers = newArgs.ForwardTo
+
+	managerOpts, err := c.getManagerOptions(newArgs)
+	if err != nil {
+		return err
+	}
+
+	switch {
+	case c.manager == nil:
+		// First call to Update; build the tailer.
+		c.manager = NewManager(c.opts.Logger, managerOpts)
```

Comment:
Since `NewManager` accepts being called with nil options, you can simplify this logic a bit by creating the manager once when the component is created.